core/src/docs/rfcs/3734_buffered_reader.md

```diff
@@ -0,0 +1,175 @@
+
+```rust
+op.reader_with(path).buffer(32 * 1024 * 1024).await
+```
+
+# Reference-level explanation
+
+This feature will be implemented in the `CompleteLayer`, with the addition of a `BufferReader` struct in `raw/oio/reader/buffer_reader.rs`. 
+
+The `BufferReader` employs` bytes::BytesMut` as the inner buffer and utilizes `(Option<usize>, Option<usize>)` to track the buffered range of the file.
+
+
+```rust
+     ...
+     async fn read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::Reader)> {
+          BufferReader::new(self.complete_read(path, args).await)
+     }
+
+     ...
+
+    fn blocking_read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::BlockingReader)> {
+          BufferReader::new(self.complete_blocking_read(path, args))
+    }
+     ...
+```
+
+A `buffer` field of type `Option<usize>` will be introduced to `OpRead`. If `buffer` is set to `None`, it functions with default behavior. However, if buffer is set to `Some(n)`, it denotes the maximum buffer capability that the `BufferReader` can utilize, and buffering behaviors introduced below.
+
+**Buffering**
+
+When the `Buffer Capability` is specified, the underlying reader fetches and buffers data in chunks corresponding to the buffer size.
+
+```
+Buffer
+┌───────┐
+│       │
+└───────┘
+File
+┌───┬───────┬──┐
+│   │       │  │
+└───▲───────▲──┘
+    │       │
+    │       │ 2. ReadToEnd(limit), The limit <= Cap(Buffer).
+    │
+    │ 1. SeekFromStart(128)
+
+```
+
+```
+Buffer
+┌───────┐
+│*******│
+└───────┘
+File
+┌───┬───────┬──┐
+│   │*******│  │
+└───▲───────▲──┘
+     3. Fetches and buffers the data.
+```
+
+**Tailing Read**
+
+If a read request attempts to read the trailing bytes of the file, it only reads and buffers the `Min(Max(Limit,Cap(Buffer)),FileEnd-Cursor)` bytes.
+
+```
+File
+┌──────────────┐
+│              │
+└────────────▲─▲
+             │ │
+             │ │ 2. ReadToEnd(limit)
+             │
+             │ 1. SeekFromEnd(-8)
+```
+
+**Overlapping Read**
+
+If a read request attempts to read data that is partially buffered, it copies the partially buffered data to the `dst` buffer first.
+
+```
+Buffer
+┌───────┐
+│*******│
+└───────┘
+File
+┌───────┬──────┐
+│*******│      │
+└────▲──┴──▲───┘
+     │     │
+     │     │ 2. ReadToEnd(limit), The limit <= Cap(Buffer).
+     │
+     │ 1. SeekFromStart(128)
+```
+
+Then, fetches and buffers data in chunks corresponding to the buffer size. Finally, it copies the another part data to the `dst` buffer.
+
+```
+Buffer
+┌───────┐
+│///////│
+└───────┘
+File
+┌───────┬──────┐
+│       │//////│
+└────▲──┴──▲───┘
+     3. Fetches and buffers the data.
+```
+
+**Bypass**
+
+If a read request attempts to read data larger than the `Buffer Capability`, it fetches the data bypassing the `Buffer` entirely.
+
+```
+Buffer
+┌───────┐
+│       │
+└───────┘
+File
+┌───────────┬──┐
+│           │**│
```

Comment:
Can you explain what the blacked-out part in the back of this confusing picture is? It seems that doesn't have any meaning in this scenario.

Reply:
Yes, I will remove it later. It's kind of confusing. Thank🥰